src/main/java/Duke.java

```diff
@@ -1,10 +1,364 @@
+            if (line.charAt(3) == 'T') {
+                tasks.add(new Todo(line.substring(9)));
+            } else if (line.charAt(3) == 'D') {
+                readDeadlineIntoTasksList(tasks, line);
+            } else if (line.charAt(3) == 'E') {
+                readEventIntoTasksList(tasks, line);
+            }
+            if (line.charAt(6) == 'X') {
+                tasks.get(index).setTaskStatus(true);
+            } else if (line.charAt(6) == ' ') {
+                tasks.get(index).setTaskStatus(false);
+            }
+            index++;
+        }
+    }
+
+    /**
+     * Reads deadline tasks into tasks list.
+     *
+     * @param tasks The ArrayList holding the task objects.
+     * @param line  The deadline to be read into tasks ArrayList.
+     */
+
+    private static void readDeadlineIntoTasksList(ArrayList<Task> tasks, String line) {
+        int byIndex = line.indexOf("(by");
+        String taskDescription = line.substring(9, byIndex);
+        String deadline = line.substring(byIndex + 4, line.length() - 1);
+        tasks.add(new Deadline(taskDescription, deadline));
+    }
+
+    /**
+     * Reads event tasks into tasks list.
+     *
+     * @param tasks The ArrayList holding the task objects.
+     * @param line  The event to be read into tasks ArrayList.
+     */
+
+    private static void readEventIntoTasksList(ArrayList<Task> tasks, String line) {
+        //index of event start date/time
+        int fromIndex = line.indexOf("(from");
+        String event = line.substring(9, fromIndex);
+        //index of event end date/time
+        int toIndex = line.indexOf("to");
+        String eventStart = line.substring(fromIndex + 6, toIndex);
+        String eventEnd = line.substring(toIndex + 3, line.length() - 1);
+        tasks.add(new Event(event, eventStart, eventEnd));
+    }
+
+    private static void printFileContents(File f) throws FileNotFoundException {
+        // create a Scanner using the File as the source
+        Scanner s = new Scanner(f);
+        while (s.hasNext()) {
+            System.out.println(s.nextLine());
+        }
+    }
+
+    private static void clearFileContents(File f) throws IOException {
+        FileWriter fw = new FileWriter(f);
+        fw.write("");
+        fw.close();
+    }
+
+    private static void appendToFile(File f, String textToAppend) throws IOException {
+        FileWriter fw = new FileWriter(f, true); // create a FileWriter in append mode
+        fw.write(textToAppend);
+        fw.close();
+    }
+
+    /**
+     * Executes duke commands: list, mark, unmark, todo, deadline, and event with exception handling.
+     *
+     * @param command The user inputted command.
+     * @param tasks   An ArrayList holding the task objects.
+     * @throws DukeException if input is not any of the supported duke commands.
+     */
+
+    private static void executeDukeCommands(String command, ArrayList<Task> tasks) throws DukeException {
```

Comment:
can try index out of bound check as guard clause and then resume the happy path.
reduces arrow head code